mlflow/java/src/main/java/org/mlflow/sagemaker/MLeapPredictor.java

```diff
@@ -1,18 +1,109 @@
 package org.mlflow.sagemaker;
 
+import com.fasterxml.jackson.core.JsonProcessingException;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import ml.combust.mleap.runtime.MleapContext;
+import ml.combust.mleap.runtime.frame.DefaultLeapFrame;
+import ml.combust.mleap.runtime.frame.Row;
 import ml.combust.mleap.runtime.frame.Transformer;
+import ml.combust.mleap.runtime.javadsl.BundleBuilder;
+import ml.combust.mleap.runtime.javadsl.ContextBuilder;
+import org.mlflow.utils.SerializationUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.collection.JavaConverters;
+import scala.collection.Seq;
 
 /** A {@link org.mlflow.sagemaker.Predictor} implementation for the MLeap model flavor */
 public class MLeapPredictor extends Predictor {
-  public MLeapPredictor(String modelDataPath, String inputSchemaPath) {}
+  private final Transformer pipelineTransformer;
+  private final LeapFrameSchema inputSchema;
+
+  private static final Seq<String> predictionColumnNames;
+  private static final Logger logger = LoggerFactory.getLogger(MLeapPredictor.class);
+
+  static {
+    List<String> predictionColumnList = Arrays.asList("prediction");
+    predictionColumnNames =
+        JavaConverters.asScalaIteratorConverter(predictionColumnList.iterator()).asScala().toSeq();
+  }
+
+  /**
+   * Constructs an {@link MLeapPredictor}
+   *
+   * @param modelDataPath The path to the serialized MLeap model
+   * @param inputSchema The path to JSON-formatted file containing the input schema that the model
+   *     accepts
+   */
+  public MLeapPredictor(String modelDataPath, String inputSchemaPath) {
+    MleapContext mleapContext = new ContextBuilder().createMleapContext();
+    BundleBuilder bundleBuilder = new BundleBuilder();
+    this.pipelineTransformer = bundleBuilder.load(new File(modelDataPath), mleapContext).root();
+    try {
+      this.inputSchema = LeapFrameSchema.fromPath(inputSchemaPath);
+    } catch (IOException e) {
+      logger.error("Could not read the model input schema from the specified path", e);
+      throw new PredictorLoadingException(String.format(
+          "Failed to load model input schema from specified path: %s", inputSchemaPath));
+    }
+  }
 
   @Override
-  protected PredictorDataWrapper predict(PredictorDataWrapper input) {
-    throw new UnsupportedOperationException("Not yet implemented!");
+  protected PredictorDataWrapper predict(PredictorDataWrapper input)
```

Comment:
OK, my understanding of the data flow is we:

  1. Take a string off the wire, which we expect is JSON.
  2. Deserialize that JSON in a manner which attempts to be analogous to if the data were a Pandas DataFrame serialized as JSON, i.e., into a list of dictionaries, one dictionary per row (or `List<Map<String, Object>>` in Java).
  3. Re-serialize the JSON into a String of similar format, but with two main differences: (A) We drop columns that are not expected from the serialized schema and (B) we add the schema as a top-level field.
  4. Ask MLeap to deserialize the JSON string into a DefaultLeapFrame.

We then ask the model to predict based on this DefaultLeapFrame, which results in a new LeapFrame. We then:
  1. Throw out the schema, pulling out just the rows, which we for some reason assume the first column of each row is the result of interest.
  2. Serialize the predictions as an array of simple types.
  3. Output that array back to the user.

Is the above flow correct?

Reply:
That flow is correct. Regarding (1) from the second portion, we only want to use rows from the column named `prediction` - this is what our `pyfunc` wrapper for Spark models does. Accordingly, the statically definitely `predictionColumnNames` list is misleading; I defined it because the `LeapFrame.select()` operation takes a list. I think it would be better to statically define the column name as a string and create a new list containing the single column name within the method that calls `LeapFrame.select()`.